docs/proposals/107-spark-client/README.md

```diff
@@ -0,0 +1,597 @@
+
+- Creating SparkApplication CRDs
+- Managing Spark Connect servers
+- Writing YAML configurations
+- Handling cleanup
+
+## Goals
+
+1. Simple Python API for Spark on Kubernetes
+2. Auto-provision Spark Connect servers
+3. Support connecting to existing servers
+4. Full PySpark compatibility
+5. Extensible architecture for future batch job support
+6. Submit and manage batch Spark jobs via SparkApplication CRD
+7. Kubeflow ecosystem integration (Pipelines, Trainer, Spark History MCP Server)
+
+## Non-Goals
+
+- Supporting Spark outside Kubernetes (local mode, standalone clusters)
+- Managing Spark Operator installation
+- Replacing the Spark Operator
+
+---
+
+## API
+
+### Basic Usage
+
+```python
+from kubeflow.spark import SparkClient
+
+# Creates Spark Connect server automatically with defaults
+with SparkClient.builder().build() as client:
+    spark = client.session()
+    df = spark.sql("SELECT * FROM my_table")
+    df.show()
+# Auto cleanup on exit
+```
+
+### Quick One-Liner
+
+```python
+from kubeflow.spark import SparkClient
+
+# Get SparkSession directly
+spark = SparkClient.builder().num_executors(5).get_or_create()
+```
+
+### Connect to Existing Server
+
+```python
+from kubeflow.spark import SparkClient
+
+# Connect to existing Spark Connect server (no auto-provisioning)
+client = SparkClient.connect("sc://spark-server:15002")
+spark = client.session()
+df = spark.read.parquet("s3a://bucket/data/")
+client.stop()
+```
+
+### Full Configuration (Builder Pattern)
+
+```python
+from kubeflow.spark import SparkClient
+
+client = (
+    SparkClient.builder()
+    .namespace("spark-jobs")
+    .driver(cores=2, memory="4g")
+    .executor(cores=4, memory="8g")
+    .num_executors(10)
+    .image("my-spark:3.5.0")
+    .spark_conf("spark.sql.shuffle.partitions", "200")
+    .spark_conf("spark.hadoop.fs.s3a.endpoint", "s3.amazonaws.com")
+    .executor_gpu("nvidia.com/gpu", 1)
+    .volume("data", "/mnt/data", persistentVolumeClaim={"claimName": "my-pvc"})
+    .build()
+)
+
+spark = client.session(app_name="my-analysis")
+# ...
+client.stop()
+```
+
+---
+
+## SparkClient API
+
+### SparkClientBuilder
+
+The Builder Pattern provides a fluent API for configuration - readable, discoverable, and extensible.
+
+```python
+class SparkClientBuilder:
+    """Builder for SparkClient configuration."""
+
+    # Kubernetes settings
+    def namespace(self, ns: str) -> "SparkClientBuilder":
+        """Set Kubernetes namespace."""
+
+    def service_account(self, sa: str) -> "SparkClientBuilder":
+        """Set service account for Spark pods."""
+
+    # Resource configuration
+    def driver(self, cores: int = 1, memory: str = "1g") -> "SparkClientBuilder":
+        """Configure driver resources."""
+
+    def executor(self, cores: int = 1, memory: str = "1g") -> "SparkClientBuilder":
+        """Configure executor resources."""
+
```

Comment:
I am missing a bit the CPU and memory requests vs Limit . I only see

``` 
.driver(cores=2, memory="4g")
.executor(cores=4, memory="8g")
```

which is somehow mixing up requests with limits. Furthermore we should make sure that it also works with the integrated seaweedfs S3.

Reply:
Thanks for pointing it out. There are lot of conf is spark so anyways we need wider key value pair set.